### ISSUE TITLE
System Exception--Illegal Instruction

### ISSUE THREAD

**Issue by tyaniw** [pinterf/mvtools#9]
All x64 versions from 2.6.0.5 up to the most recent 2.7... fail due to MAnalyse.  This is on Windows 7 64 bit, Phenom II X4 945.  Please advise?

**Comment by pinterf** [comment]
Question 1: 
For reproducing the issue, I need more info on the parameters used, could you please provide your script?
Question 2: which Avisynth+ version are you using?
Question 3: Perhaps you should run avsmeter utility (avsmeter64 -avsinfo), is it telling everything is OK?

**Comment by tyaniw** [comment]
1:
I've attached the simple script I've been using for testing at the bottom.  MSuper works just fine and displays the multilevel clip in Virtualdub--it is MAnalyse that causes the "illegal instruction"
The script also works in 32-bit Avisynth 2.6 with MVTools 2.6.0.5 x86.

2:
Avisynth+ r2508MT x64

3:
AVSMeter simply tells me the problem ("illegal instruction") I already saw when attempting to open the script in Virtualdub (64 bit).
The log didn't seem to indicate that I was missing any dependencies.

mpeg2source("E:\box set\dogfights\0 The Greatest Air Battles.d2v")
assumetff()

tfm()

sup1= Dither_Luma_Rebuild(S0=1.0,c=0.0625,uv=1).MSuper(pel=2,sharp=2,hpad=16,vpad=16)
sup2= MSuper(pel=2,sharp=2,hpad=16,vpad=16)
bv1 = MAnalyse(sup1,isse=false,isb=true,delta=1,blksize=16,overlap=8)
fv1    = MAnalyse(sup1,isse=false,isb=false,delta=1, blksize=16,overlap=8)
MDegrain1(sup2,bv1,fv1,thsad=400)

**Comment by tyaniw** [comment]
I should also add that I tried using several blocksizes (8, 16, 32) with various overlap values, and tried changing pel to 1.  No effect...

**Comment by DJATOM** [comment]
Intel® Core™2 Quad Q6600 has a similar problem. x86 works fine, x64 gives "illegal instruction". I don't know why.

**Comment by pinterf** [comment]
Thanks.

I emulated a Core2 Merom architecture with Intel SDE emulator and got this error:

```
TID 0 SDE-ERROR: Executed instruction not valid for specified chip (CORE2): 0x7fedbcbb9f9: pcmpeqq xmm0, xmmword ptr [rsp+0x20]
Image: C:\Program Files (x86)\AviSynth+\plugins64\mvtools2.dll+0x13b9f9
Function: conc::Interlocked::Data128::operator==
File Name: c:\github\mvtools\sources\conc\interlocked.hpp:168
Instruction bytes are: 66 0f 38 29 44 24 20
```

The problem is with _mm_cmpeq_epi64, which is an SSE4 instruction.

Problematic lines in interlocked.hpp, which are active only in x64 builds:

```
bool	Interlocked::Data128::operator == (const Data128 & other) const
{
	__m128i var1 = _mm_load_si128((__m128i*)_data);
	__m128i var2 = _mm_load_si128((__m128i*)other._data);
	__m128i result = _mm_cmpeq_epi64(var1, var2);
	result = _mm_and_si128(result, _mm_srli_si128(result, 8));
	return !!_mm_cvtsi128_si32(result);
}
```

I'll prepare the fix.

**Comment by pinterf** [comment]
Fixed in 2.7.22

**Comment by tyaniw** [comment]
Excellent, thanks!
